File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        equiv = {
            '(': ')',
            '[': ']',
            '{': '}'
        }
        stack = []

        for char in s:
            if char in ['(', '[', '{']:
                stack.append(equiv[char])
            else:
                if len(stack) == 0:
                    return False
                if char == stack[-1]:
                    stack.pop(-1)
                else:
                    return False
        
        if len(stack) > 0:
            return False

        return True

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        var closingPars = List.of(')', '}', ']');
        if (closingPars.contains(s.charAt(0))) return false;
        
        Stack<Character> openStack = new Stack<Character>();
        openStack.push(s.charAt(0));
        


        for(var p: s.substring(1).toCharArray()){
            if (closingPars.contains(p)){
                if (openStack.empty()) return false;
                Character head = (Character) openStack.peek();
                if (p==(')') && !head.equals('(') ) return false;
                if (p==('}') && !head.equals('{') ) return false;
                if (p==(']') && !head.equals('[') ) return false;
                if (p==(')') || p==('}') || p==(']') ) openStack.pop();
            }
            else openStack.push(p);
            

        }
    if (!openStack.empty()) return false; else return true;
    }
}